ddl/column.go

```diff
@@ -427,13 +427,118 @@ func (d *ddl) onSetDefaultValue(t *meta.Meta, job *model.Job) error {
 func (d *ddl) onModifyColumn(t *meta.Meta, job *model.Job) error {
 	newCol := &model.ColumnInfo{}
 	oldColName := &model.CIStr{}
-	err := job.DecodeArgs(newCol, oldColName)
+	pos := &ast.ColumnPosition{}
+	err := job.DecodeArgs(newCol, oldColName, pos)
 	if err != nil {
 		job.State = model.JobCancelled
 		return errors.Trace(err)
 	}
 
-	return errors.Trace(d.updateColumn(t, job, newCol, oldColName))
+	if pos.Tp == ast.ColumnPositionNone {
+		return errors.Trace(d.updateColumn(t, job, newCol, oldColName))
+	}
+
+	tblInfo, err := t.GetTable(job.SchemaID, job.TableID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	col := findCol(tblInfo.Columns, newCol.Name.L)
+	if col == nil {
+		col = findCol(tblInfo.Columns, oldColName.L)
+	}
+
+	originalState := col.State
+	switch col.State {
+	case model.StatePublic:
+		if err = d.modifyColumnPosition(tblInfo, newCol, oldColName, pos); err != nil {
+			job.State = model.JobCancelled
+			return errors.Trace(err)
+		}
+		// use delete only to prevent add record based on different column order.
+		// otherwise `insert into t value (1, 2)` will add different row and index record.
+		newCol.State = model.StateDeleteOnly
```

Comment:
```SQL
create table t1 (a int, b int);
insert into t1 value (1, 2);
alter table t1 modify column b int first;
insert into t1 value (1, 2);
```
```
mysql> select * from t1;
+------+------+
| b    | a    |
+------+------+
|    2 |    1 |
|    1 |    2 |
+------+------+
2 rows in set (0.00 sec)
```
As you can see, `insert into t1 value (1, 2)` have different meaning in different schema, I know that both of them are right, are persistent, are consistent, and can be read at any time in the future.

But when user see `alter table t1 modify column b int first` returned successfully, he want to see all node in cluster have changed the column order, this means all subsequent `insert into t1 value (1, 2)` must explain as `b = 1 a = 2`.

F1 allow two schema in use at same time. So if we change order directly and return immediately, user will think `all schema in all node have changed, so I can do “insert into t1 value (1, 2)” to add a record (b = 1, a = 2)` but the fact is some node contain old schema which means (a = 1, b = 2).

This may not be a big problem, but I have to ask your opinion. @zimulala